### PR TITLE
Adding feature to allow signed assertions to be set via delegate

### PR DESCRIPTION
**Pull request by trwalke**
fix for https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2184

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs [hunk 1]
```diff
@@ -152,6 +152,24 @@ public ConfidentialClientApplicationBuilder WithClientAssertion(string signedCli
             return this;
         }
 
+        /// <summary>
+        /// Sets the application client assertion. See https://aka.ms/msal-net-client-assertion
+        /// </summary>
+        /// <param name="clientAssertionDelegate">The client assertion delegete used to prove the identity of the application to Azure AD.
```
Comment (jmprieur) [inline src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:158]:
```suggestion
        /// <param name="clientAssertionDelegate">delegate computing the client assertion used to prove the identity of the application to Azure AD.
```

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs [hunk 2]
```diff
@@ -152,6 +152,24 @@ public ConfidentialClientApplicationBuilder WithClientAssertion(string signedCli
...
+        /// This is a delegete that computes a Base-64 encoded JWT for each authentication call.</param>
```
Comment (jmprieur) [inline src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:159]:
```suggestion
        /// This is a delegate that computes a Base-64 encoded JWT for each authentication call.</param>
```

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs [hunk 3]
```diff
@@ -152,6 +152,24 @@ public ConfidentialClientApplicationBuilder WithClientAssertion(string signedCli
...
+        /// <returns></returns>
```
Comment (jmprieur) [inline src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:160]:
```suggestion
        /// <returns>The ConfidentialClientApplicationBuilder to chain more .With methods</returns>
```

**Review by jmprieur — APPROVED** [review]
LGTM
I left a few suggestions/typo fixes for the XML comment

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs [hunk 4]
```diff
@@ -548,7 +554,35 @@ public async Task ConfidentialClientUsingSignedClientAssertionTestAsync()
...
+                Assert.AreEqual(TestConstants.s_scope.AsSingleString(), result.Scopes.AsSingleString());
+
+                // make sure user token cache is empty
+                Assert.AreEqual(0, app.UserTokenCacheInternal.Accessor.GetAllAccessTokens().Count());
+                Assert.AreEqual(0, app.UserTokenCacheInternal.Accessor.GetAllRefreshTokens().Count());
 
+                // check app token cache count to be 1
+                Assert.AreEqual(1, app.AppTokenCacheInternal.Accessor.GetAllAccessTokens().Count());
+                Assert.AreEqual(0, app.AppTokenCacheInternal.Accessor.GetAllRefreshTokens().Count()); // no RTs are returned
+
```
Comment (jennyf19) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs:582]: why are there no rts?

**Review by jennyf19 — APPROVED** [review]
:shipit:

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs (hunk 4 above)
Comment (bgavrilMS) [reply tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs:582]: I think it's because Travis is using `AcquireTokenForClient` to test this grant. Only ATs are generated in this case.

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs [hunk 5]
```diff
@@ -548,7 +554,35 @@ public async Task ConfidentialClientUsingSignedClientAssertionTestAsync()
...
+                // assert client credential
+                Assert.IsNotNull(app.ClientCredential.SignedAssertion);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs:584]: Did you add extra assertions after the test was passing? This is actually null, `app.ClientCredential.SignedAssertionDelegate` is used

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs [hunk 6]
```diff
@@ -152,6 +152,24 @@ public ConfidentialClientApplicationBuilder WithClientAssertion(string signedCli
             return this;
         }
 
+        /// <summary>
+        /// Sets the application client assertion. See https://aka.ms/msal-net-client-assertion
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:156]:
```suggestion
        /// Configures a delegate that creates a client assertion. See https://aka.ms/msal-net-client-assertion
```

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs [hunk 7]
```diff
@@ -152,6 +152,24 @@ public ConfidentialClientApplicationBuilder WithClientAssertion(string signedCli
...
+        /// </summary>
+        /// <param name="clientAssertionDelegate">delegate computing the client assertion used to prove the identity of the application to Azure AD.
+        /// This is a delegate that computes a Base-64 encoded JWT for each authentication call.</param>
+        /// <returns>The ConfidentialClientApplicationBuilder to chain more .With methods</returns>
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:160]:
```suggestion
        /// <returns>The ConfidentialClientApplicationBuilder to chain more .With methods</returns>
        /// <remarks> Callers can use this mechanism to cache their assertions </remarks>
```

File: src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs [hunk 8]
```diff
@@ -152,6 +152,24 @@ public ConfidentialClientApplicationBuilder WithClientAssertion(string signedCli
...
+        public ConfidentialClientApplicationBuilder WithClientAssertion(Func<string> clientAssertionDelegate)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/ConfidentialClientApplicationBuilder.cs:161]: I would add a comment on the old `WithClientAssertion(string)` as well to tell users that if they want to cache their assertions, they can use this method.

**Review by bgavrilMS — APPROVED** [review]
LGTM, please fix the unit test and consider the public api comment changes.

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs (hunk 5 above)
Comment (trwalke) [reply tests/Microsoft.Identity.Test.Unit/PublicApiTests/ConfidentialClientApplicationTests.cs:584]: no, this was a typo from pressing control z a few too many times. Test is working

**Review by bgavrilMS — APPROVED** [review]

**Comment by bgavrilMS** [comment]
@trwalke - this looks good Travis, please merge it.
